src/components/LoginForm.js

```diff
@@ -2,80 +2,150 @@ import React from 'react';
+                  </div>
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+      </LoginForm>
+    );
+  }
+}
+
 @ReactMixin.decorate(History)
 export default class LoginForm extends React.Component {
   state = {
-    username: '',
-    password: '',
-    isProcessing: false,
-    errorMessage: null
+    fields: {
+      username: '',
+      password: '',
+    },
+    errorMessage: null,
+    isFormProcessing: false
   };
 
   onFormSubmit(e) {
     e.preventDefault();
 
-    var self = this;
+    var next = (err, data) => {
+      if (err) {
+        return this.setState({
+          isFormProcessing: false,
+          errorMessage: err.message
+        });
+      }
 
-    var redirectTo = this.props.redirectTo || '/';
+      // If the user didn't specify any data,
+      // then simply default to what we have in state.
+      data = data || this.state.fields;
 
-    self.setState({ isProcessing: true });
+      UserActions.login({
+        username: data.username,
+        password: data.password
+      }, (err, result) => {
+        this.setState({ isFormProcessing: false });
 
-    UserActions.login({
-      username: this.state.username,
-      password: this.state.password
-    }, function (err, result) {
-      self.setState({ isProcessing: false });
+        if (err) {
+          this.setState({ errorMessage: err.message });
+        } else {
+          this.history.pushState(null, this.props.redirectTo || '/');
+        }
+      });
+    };
 
-      if (err) {
-        self.setState({ errorMessage: err.message });
-      } else {
-        self.history.pushState(null, redirectTo);
-      }
+    this.setState({
+      isFormProcessing: true
     });
+
+    if (this.props.onSubmit) {
+      e.data = this.state.fields;
+      this.props.onSubmit(e, next);
+    } else {
+      next(null, this.state.fields);
+    }
   }
 
-  onUsernameChanged(e) {
-    this.setState({ username: e.target.value });
+  _mapFormFieldHandler(element, tryMapField) {
+    if (['input', 'textarea'].indexOf(element.type) > -1) {
+      if (element.props.type !== 'submit') {
+        switch(element.props.name) {
+          case 'login':
+          case 'username':
+            tryMapField('username');
+            break;
+          case 'password':
+            tryMapField('password');
+            break;
+        }
+      }
+    }
   }
 
-  onPasswordChanged(e) {
-    this.setState({ password: e.target.value });
+  _spIfHandler(action, element) {
+    var test = null;
+
+    switch (action) {
+      case 'form.processing':
+        test = this.state.isFormProcessing;
+        break;
+      case 'form.error':
+        test = !!this.state.errorMessage;
+        break;
+    }
+
+    return test;
+  }
+
+  _spBindHandler(action, element) {
+    var element = false;
+
+    switch (action) {
+      case 'form.errorMessage':
+        var className = element.props ? element.props.className : undefined;
```

Comment:
Since this is a block variable, isn't it better to use [`let`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/let) instead of `var`?